### PR TITLE
MicroProfile Config 3.0

### LINKED ISSUE #2640: Config 2.0
- [ ] Config 2.0 support
- [ ] Documentation
- [ ] Samples
- [ ] Guides

MicroProfile 4.0 support epic: #2636

### PR DESCRIPTION
**Pull request by tomas-langer**
Implementation of MP Config
Enabled TCK for Config
Fixed all related tests

Resolves #2640

### CODE REVIEW

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java [hunk 1]
```diff
@@ -107,11 +106,15 @@
...
     @Override
-    public ConfigValue getConfigValue(String s) {
-        // TODO requires implementation of latest config version
-        return null;
+    public ConfigValue getConfigValue(String key) {
+        if (configProfile == null) {
+            return findConfigValue(key)
+                    .orElseGet(() -> new ConfigValueImpl(key, null, null, null, 0));
+        }
+        return findConfigValue("%" + configProfile + "." + key)
```
Comment (spericas) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:115]: Is the need for all this String concatenation required by the spec? Seems like not the fastest option

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java [hunk 2]
```diff
@@ -365,32 +386,70 @@ private String resolveReferences(String key, String value) {
...
         try {
-            return format(value);
-        } catch (NoSuchElementException e) {
-            LOGGER.log(Level.FINER, e, () -> String.format("Reference for key %s not found. Value: %s", key, value));
-            return value;
+            if (value.contains("${")) {
+                return processExpressions(value);
+            } else {
+                return value;
+            }
```
Comment (spericas) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:396]: You don't seem to like conditional expressions :)

File: microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    TypedField field1() {
+        return field1;
+    }
+
+    TypedField field2() {
+        return field2;
+    }
+
+    static final class TypedField {
```
Comment (spericas) [inline microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java:114]: Could this be a record too?

File: microprofile/config/src/main/java/io/helidon/microprofile/config/DefaultConfigValue.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.microprofile.config;
+
+import org.eclipse.microprofile.config.ConfigValue;
+
+class DefaultConfigValue implements ConfigValue {
```
Comment (spericas) [inline microprofile/config/src/main/java/io/helidon/microprofile/config/DefaultConfigValue.java:20]: Record?

**Review by spericas — APPROVED** [review]
Left some minor comments but otherwise LGTM.

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java (hunk 2 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:396]: Yes. I always found them trickier to read.

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java (hunk 1 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:115]: The specification for profile requires that you first attempt to find the configuration value by the config profile prefix (which implies string concatenation), and then (if not found) you use the original key.

Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:115]: Caching is not an option, as MP Config requires the values to be always requested from config source(s)

File: microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java (hunk 3 above)
Comment (tomas-langer) [reply microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java:114]: This is an existing class extracted to upper level. The class was not modified other than changing visibility to work with an extracted class.

File: microprofile/config/src/main/java/io/helidon/microprofile/config/DefaultConfigValue.java (hunk 4 above)
Comment (tomas-langer) [reply microprofile/config/src/main/java/io/helidon/microprofile/config/DefaultConfigValue.java:20]: Fixed

File: microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java (hunk 3 above)
Comment (tomas-langer) [reply microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java:114]: Fixed (as this removes a lot of code)

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java (hunk 2 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java:396]: Nevertheless fixed, as this one is really simple

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfigBuilder.java [hunk 5]
```diff
@@ -492,7 +498,12 @@ public Config build() {
 
         // nope, return the result
         if (configuredProfile == null) {
+            LOGGER.fine("Built MP config with no profile");
```
Comment (danielkec) [inline config/config-mp/src/main/java/io/helidon/config/mp/MpConfigBuilder.java:501]: `if (LOGGER.isLoggable(Level.FINEST)) {`

Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfigBuilder.java:501]: Not needed (also it is fine not finest). If the string is a constant, there is not much overhead doing it this way. The guarding of logs statements should only be used when passing variables to output

**Review by spericas — APPROVED** [review]

**Pull request merged**